AudioDeviceExample/AudioDevices/ExampleAudioEngineDevice.m

```diff
@@ -0,0 +1,741 @@
+
+        self.renderingContext = malloc(sizeof(RendererAudioContext));
+        memset(self.renderingContext, 0, sizeof(RendererAudioContext));
+
+        self.renderingContext->deviceContext = context;
+        self.renderingContext->maxFramesPerBuffer = _renderingFormat.framesPerBuffer;
+        if (@available(iOS 11.0, *)) {
+            self.renderingContext->renderBlock = (__bridge void *)(_engine.manualRenderingBlock);
+        }
+
+        [self setupAudioEngine];
+
+        if (self.capturingContext) {
+            if (![self setupAudioUnitWithRenderContext:self.renderingContext
+                                        capturingContext:self.capturingContext]) {
+                free(self.renderingContext);
+                self.renderingContext = NULL;
+                return NO;
+            }
+            return [self startAudioUnit];
+        }
+    }
+    return YES;
+}
+
+- (BOOL)stopRendering {
+    _rendererInitialized = NO;
+    [self stopAudioUnit];
+
+    @synchronized(self) {
+        [self teardownAudioUnit];
+        [self teardownAudioEngine];
+
+        NSAssert(self.renderingContext != NULL, @"Should have a rendering context.");
+        free(self.renderingContext);
+        self.renderingContext = NULL;
+
+        NSAssert(self.capturingContext != NULL, @"Should have a capture context.");
+        free(self.capturingContext);
+        self.capturingContext = NULL;
+    }
+
+    return YES;
+}
+
+#pragma mark - TVIAudioDeviceCapturer
+
+- (nullable TVIAudioFormat *)captureFormat {
+    if (!_capturingFormat) {
+
+        if (!self.isRendererInitialized) {
+            [self setupAVAudioSession];
+            _capturerInitialized = YES;
+        }
+
+        /*
+         * Assume that the AVAudioSession has already been configured and started and that the values
+         * for sampleRate and IOBufferDuration are final.
+         */
+        _capturingFormat = [[self class] activeFormat];
+    }
+
+    return _capturingFormat;
+}
+
+- (BOOL)initializeCapturer {
+    _audioBufferList.mNumberBuffers = 1;
+    _audioBufferList.mBuffers[0].mNumberChannels = kPreferredNumberOfChannels;
+
+    return YES;
+}
+
+- (BOOL)startCapturing:(nonnull TVIAudioDeviceContext)context {
+    @synchronized (self) {
+        self.capturingContext = malloc(sizeof(CapturerAudioContext));
+        memset(self.capturingContext, 0, sizeof(CapturerAudioContext));
+
+        self.capturingContext->deviceContext = context;
+        self.capturingContext->bufferList = &_audioBufferList;
+
+        if (self.renderingContext) {
+            if (![self setupAudioUnitWithRenderContext:self.renderingContext
+                                        capturingContext:self.capturingContext]) {
+                free(self.renderingContext);
+                self.renderingContext = NULL;
+                return NO;
+            }
+            return [self startAudioUnit];
+        }
+    }
```

Comment:
It looks like the AVAudioEngine will be forced to start rendering in order for capture to work. Isn't there a chance that the other party is not publishing a Track (say by using ExampleCoreAudioDevice), at which point this AudioDevice will not work?